### PR TITLE
[Overload] Active downstream connections resource monitor

### PR DESCRIPTION
**Pull request by nezdolik**
Signed-off-by: Kateryna Nezdolii [nezdolik@spotify.com](mailto:nezdolik@spotify.com)

Thread safe version of Active downstream connections resource monitor based on new proactive checks in overload manager framework. Continuation of work started in this https://github.com/envoyproxy/envoy/pull/15707. After this PR is merged, we could replace existing global&per-listener connections tracking mechanism in TCP Listener and plug overload manager downstream connections resource monitor instead.
-->

Commit Message:
Additional Description:
Risk Level: Low (new extension not wired up with existing code)
Testing: Done
Docs Changes: 
Release Notes:
Fixes https://github.com/envoyproxy/envoy/issues/12419

### CODE REVIEW

File: test/extensions/resource_monitors/downstream_connections/downstream_connections_monitor_test.cc [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  envoy::extensions::resource_monitors::downstream_connections::v3::DownstreamConnectionsConfig
+      config;
+  config.set_max_active_downstream_connections(1);
+  initialize(config);
+  EXPECT_EQ(0, monitor_->currentResourceUsage());
+  EXPECT_EQ(1, monitor_->maxResourceUsage());
+  EXPECT_TRUE(monitor_->tryAllocateResource(1));
+  EXPECT_FALSE(monitor_->tryAllocateResource(1));
+  EXPECT_TRUE(monitor_->tryDeallocateResource(1));
+  EXPECT_EQ(0, monitor_->currentResourceUsage());
```
Comment (KBaichoo) [inline test/extensions/resource_monitors/downstream_connections/downstream_connections_monitor_test.cc:57]: seems like we test alloc at max, but not dealloc while at min? Can we add that here -- seems like it'd fit what the test case name describes

**Review by KBaichoo — COMMENTED** [review]
Thanks for working on this! Here's a first pass

File: source/extensions/resource_monitors/downstream_connections/downstream_connections_monitor.cc [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+namespace DownstreamConnections {
+
+ActiveDownstreamConnectionsResourceMonitor::ActiveDownstreamConnectionsResourceMonitor(
+    const envoy::extensions::resource_monitors::downstream_connections::v3::
+        DownstreamConnectionsConfig& config)
+    : max_(config.max_active_downstream_connections()), current_(0){};
+
+bool ActiveDownstreamConnectionsResourceMonitor::tryAllocateResource(int64_t increment) {
+  auto current = current_.load(std::memory_order_relaxed);
+  while (current + increment <= max_) {
```
[the remark this answers is not in the document]
Comment (nezdolik) [reply source/extensions/resource_monitors/downstream_connections/downstream_connections_monitor.cc:19]: sure

File: test/extensions/resource_monitors/downstream_connections/downstream_connections_monitor_test.cc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+namespace Envoy {
+namespace Extensions {
+namespace ResourceMonitors {
+namespace DownstreamConnections {
+
+class ActiveDownstreamConnectionsMonitorTest : public testing::Test {
+public:
+  void initialize(const envoy::extensions::resource_monitors::downstream_connections::v3::
+                      DownstreamConnectionsConfig& config) {
+    monitor_ = std::make_unique<ActiveDownstreamConnectionsResourceMonitor>(config);
```
[the remark this answers is not in the document]
Comment (nezdolik) [reply test/extensions/resource_monitors/downstream_connections/downstream_connections_monitor_test.cc:18]: monitor creation through factory is covered by [config_test](test/extensions/resource_monitors/downstream_connections/config_test.cc). I followed the pattern of test layout for existing monitors, one test for config & factory and the other one for monitor itself.

**Comment by nezdolik** [comment]
@KBaichoo applied review comments. Still need to figure out how to fix failing doc check for hidden extension, it's unfortunate that it also blocks all the other checks from running.

**Comment by adisuissa** [comment]
> @KBaichoo applied review comments. Still need to figure out how to fix failing doc check for hidden extension, it's unfortunate that it also blocks all the other checks from running.

I think you can surround it with a `[#comment` tag, similar to [this](https://github.com/envoyproxy/envoy/blob/7a4bbb3c5b52b84625ad0626f916be1c99399ba8/api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto#L19) (just need to remember to remove it eventually ;))

**Review by KBaichoo — COMMENTED** [review]
Thanks for adding this, it looks great (pending CI). Can we get a release note?

**Review by KBaichoo — COMMENTED** [review]
/wait

Seems like CI still running into issues. Can we have release notes / docs for this @nezdolik ? Thank you

**Comment by nezdolik** [comment]
Thanks for the suggestion @adisuissa. I tried to replicate the configuration that linked above quic extension is using (comment out extension tag in extension proto+no entry in extensions_metadata.yaml), but CI check failed with: `CodeChecker ERROR [extensions_registered] Metadata missing for extension: envoy.resource_monitors.downstream_connections`.

**Comment by nezdolik** [comment]
@KBaichoo should we add release notes & docs at this point, as they are user facing? The extension code is there but monitor is not yet wired with code of tcp listener. It would be wired in next PR. Adding docs/release notes may be confusing for users, they may try to enable the monitor and nothing will happen in terms of limiting the number of downstream conns.

**Comment by phlax** [comment]
> do you have any suggestions on path forward?

beyond noting that the `not-implemented-hide` doesnt seem to be there - not immediate - i thought that you had figured a resolution - but will look again early next week if not

**Comment by nezdolik** [comment]
@phlax wdyt about updating the check not to require category in extensions_metadata.yaml for extensions which are commented out? Eg like this [one](https://github.com/envoyproxy/envoy/pull/23361/files#diff-ef33016bf1409ea307759e52dcb4cfd956627fdcb162a4718ad5de389ca60f9eR15).
On main branch we have such [extension](https://github.com/envoyproxy/envoy/blob/7a4bbb3c5b52b84625ad0626f916be1c99399ba8/api/envoy/extensions/transport_sockets/quic/v3/quic_transport.proto#L19) with no category registered in extensions_metadata.yaml.

**Comment by phlax** [comment]
> wdyt about updating the check not to require category in extensions_metadata.yaml for extensions which are commented out? Eg ...

almost certainly a good idea - i didnt get to looking at the check specifics yet - but i think we need some way at least for wip/alpha to skip these checks and/or not be included

**Comment by phlax** [comment]
> On main branch we have such ...

hmm, im wondering how that one does it

**Comment by KBaichoo** [comment]
@nezdolik , whoops you're right, makes sense to wait for it to be fully wired

waiting for docs fix.

/wait

**Comment by phlax** [comment]
> waiting for docs fix.

apologies for delay on my side - ill look at this first thing next week

**Comment by nezdolik** [comment]
after discussion with @phlax on how to overcome CI issue, we agreed to add docs for extension and clearly state that extension is wip and not yet usable.

**Comment by phlax** [comment]
in addition to any information about the status of the extension that you might want to add - dont forget to add:

```protobuf
 option (xds.annotations.v3.file_status).work_in_progress = true;
```

... to the proto file - this will add a standard warning to the docs

you might need to add:

```starlark
    "@com_github_cncf_udpa//xds/annotations/v3:pkg",
```
to the `BUILD` file deps also

we should probably add to the docs somewhere about adding an extension with WIP code (and/or add to what is there already)

**Comment by nezdolik** [comment]
@phlax created [issue](https://github.com/envoyproxy/envoy/issues/23732) for developer docs.

**Comment by nezdolik** [comment]
Have unhidden the docs for extension:
Screenshot 2022-10-28 at 10 16 54

**Comment by nezdolik** [comment]
@KBaichoo this is ready for review.

**Review by KBaichoo — APPROVED** [review]
nice work, thank you for adding this!

**Comment by KBaichoo** [comment]
ping @markdroth re api shepard review

**Pull request merged**